latch/ldata/path.py

```diff
@@ -0,0 +1,272 @@
+}
+
+
+@dataclass
+class _Cache:
+    """Internal cache class to organize information for a `LPath`."""
+
+    path: Optional[str] = None
+    node_id: Optional[str] = None
+    name: Optional[str] = None
+    type: Optional[LDataNodeType] = None
+    size: Optional[int] = None
+    content_type: Optional[str] = None
+
+
+@dataclass
+class LPath:
+
+    _cache: _Cache = field(
+        repr=False,
+        hash=False,
+        compare=False,
+    )
+
+    path: str
+
+    def __init__(self, path: Union[str, "LPath"]):
+        if isinstance(path, LPath):
+            path = path.path
+        if not path.startswith("latch://"):
+            raise ValueError(f"Invalid LPath: {path} is not a Latch path")
+        self.path = path
+        self._cache = _Cache()
+        self._download_idx = 0
+
+    def load(self) -> None:
+        """(Re-)populate this LPath's instance's cache.
+
+        Future calls to most getters will return immediately without making a network request.
+
+        Always makes a network request.
+        """
+        data = execute(
+            gql.gql("""
+            query GetNodeData($path: String!) {
+                ldataResolvePathToNode(path: $path) {
+                    path
+                    ldataNode {
+                        finalLinkTarget {
+                            id
+                            name
+                            type
+                            removed
+                            ldataObjectMeta {
+                                contentSize
+                                contentType
+                            }
+                        }
+                    }
+                }
+            }"""),
+            {"path": self.path},
+        )["ldataResolvePathToNode"]
+
+        if data is None or data["ldataNode"] is None:
+            raise FileNotFoundError(f"No such Latch file or directory: {self.path}")
+
+        self._cache.path = self.path
+
+        final_link_target = data["ldataNode"]["finalLinkTarget"]
+        self._cache.node_id = final_link_target["id"]
+        self._cache.name = final_link_target["name"]
+        self._cache.type = LDataNodeType(final_link_target["type"].lower())
+
+        meta = final_link_target["ldataObjectMeta"]
+        if meta is not None:
+            self._cache.size = (
+                -1
+                if meta["contentSize"] is None
+                else int(final_link_target["ldataObjectMeta"]["contentSize"])
+            )
+            self._cache.content_type = final_link_target["ldataObjectMeta"][
+                "contentType"
+            ]
+
+    def node_id(self, *, load_if_missing: bool = True) -> str:
+        match = node_id_regex.match(self.path)
+        if match:
+            return match.group("id")
+
+        if self._cache.node_id is None or self._cache.path != self.path:
+            if not load_if_missing:
+                return None
+            self.load()
+        return self._cache.node_id
```

Comment:
```suggestion
        if load_if_missing and (self._cache.node_id is None or self._cache.path != self.path):
            self.load()
            
        return self._cache.node_id
```
hyper minor nit

Reply:
i didn't do it this way because this could lead to stale data being returned in the following case:

```
x = LData('latch://file.txt')
x.node_id() # returns 12345
x.path = 'latch://other.txt' # this is definitely kinda weird for user to do but still possible
x.name(load_if_missing=False) # returns 12345 -- which is incorrect (I think it makes sense to return None here instead of the node id of the previous file)
```